// js-restful/src/js-restful.js
"use strict";
var decorators_1 = require('./core/decorators');
exports.GET = decorators_1.GET;
exports.POST = decorators_1.POST;
exports.PUT = decorators_1.PUT;
exports.DELETE = decorators_1.DELETE;
exports.Path = decorators_1.Path;
exports.PathParam = decorators_1.PathParam;
exports.HeaderParam = decorators_1.HeaderParam;
exports.QueryParam = decorators_1.QueryParam;
exports.Context = decorators_1.Context;
exports.PermitAll = decorators_1.PermitAll;
exports.SecurityContext = decorators_1.SecurityContext;
exports.RolesAllowed = decorators_1.RolesAllowed;
var decorator_parser_1 = require('./core/decorator-parser');
exports.ServiceParser = decorator_parser_1.ServiceParser;
var descriptions_1 = require('./core/descriptions');
exports.ServiceDescription = descriptions_1.ServiceDescription;
exports.MethodDescription = descriptions_1.MethodDescription;
exports.HttpMethod = descriptions_1.HttpMethod;
exports.ParamDescription = descriptions_1.ParamDescription;
exports.ContextTypes = descriptions_1.ContextTypes;
var decoratorUtil_1 = require('./core/decoratorUtil');
exports.DecoratorUtil = decoratorUtil_1.DecoratorUtil;
