// js-restful/src/core/test-classes.spec.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var decorators_1 = require('./decorators');
var descriptions_1 = require('./descriptions');
var Book = (function () {
    function Book() {
    }
    return Book;
}());
exports.Book = Book;
var BookService = (function () {
    function BookService() {
        this.testProperty = 2;
    }
    BookService.prototype.unDecoratedMethod = function () { };
    BookService.prototype.allBooks = function () {
        return [];
    };
    BookService.prototype.createBook = function (name, token) {
        return { id: 1, name: name };
    };
    BookService.prototype.updateBook = function (id, name) {
        return { id: id, name: name };
    };
    BookService.prototype.deleteBook = function (id, time) {
        return true;
    };
    BookService.prototype.contextTest = function (req, res) {
        return true;
    };
    BookService.prototype.securityContextTest = function (context) {
        return true;
    };
    __decorate([
        decorators_1.GET(),
        decorators_1.RolesAllowed(['admin', 'user']), 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', []), 
        __metadata('design:returntype', Array)
    ], BookService.prototype, "allBooks", null);
    __decorate([
        decorators_1.Path('/:name'),
        decorators_1.POST(),
        decorators_1.PermitAll(),
        __param(0, decorators_1.PathParam('name')),
        __param(1, decorators_1.HeaderParam('token')), 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [String, String]), 
        __metadata('design:returntype', Book)
    ], BookService.prototype, "createBook", null);
    __decorate([
        decorators_1.Path('/:id/:name'),
        decorators_1.PUT(),
        __param(0, decorators_1.PathParam('id')),
        __param(1, decorators_1.PathParam('name')), 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [Number, String]), 
        __metadata('design:returntype', Book)
    ], BookService.prototype, "updateBook", null);
    __decorate([
        decorators_1.Path('/:id'),
        decorators_1.DELETE(),
        __param(0, decorators_1.PathParam('id')),
        __param(1, decorators_1.QueryParam('time')), 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [Number, Number]), 
        __metadata('design:returntype', Boolean)
    ], BookService.prototype, "deleteBook", null);
    __decorate([
        decorators_1.GET(),
        decorators_1.Path('/contexttest'),
        __param(0, decorators_1.Context(descriptions_1.ContextTypes.HttpRequest)),
        __param(1, decorators_1.Context(descriptions_1.ContextTypes.HttpResponse)), 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [Object, Object]), 
        __metadata('design:returntype', void 0)
    ], BookService.prototype, "contextTest", null);
    __decorate([
        decorators_1.GET(),
        decorators_1.Path('/securitycontext'),
        __param(0, decorators_1.SecurityContext()), 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [Object]), 
        __metadata('design:returntype', void 0)
    ], BookService.prototype, "securityContextTest", null);
    BookService = __decorate([
        decorators_1.PermitAll(),
        decorators_1.Path('/books'), 
        __metadata('design:paramtypes', [])
    ], BookService);
    return BookService;
}());
exports.BookService = BookService;
